### DISCUSSION TITLE
Weird "relationship does not exist" error on after insert trigger

### DISCUSSION THREAD

**Discussion by TQuantab** [supabase/supabase#20391] [Questions] [answered]
Hey there =)
im currently developing an invite system, and im getting a rather weird error/bug, when trying to insert into my group_users table.
The triggerfunction is the following:

```js
CREATE OR REPLACE FUNCTION create_user_preferences_entry()
RETURNS TRIGGER AS $$
DECLARE
    invite RECORD;
    last_group_id uuid;
BEGIN
    -- Loop through each accepted group invite for the new user
    FOR invite IN SELECT * FROM public.group_invites 
                  WHERE email = NEW.email AND accepted = true
    LOOP
        RAISE LOG 'ADDING INTO GROUP: %', invite.group_id;
        RAISE LOG 'ADDING USER: %', NEW.id;
        -- Add user to each group in group_users table
        INSERT INTO public.group_users (group_id, user_id, role)
        VALUES (invite.group_id, NEW.id, 'member');

        -- Keep track of the last group_id
        last_group_id := invite.group_id;
    END LOOP;

    -- If no accepted invites, create a new group
    IF last_group_id IS NULL THEN
        INSERT INTO public.groups (name) VALUES ('personal')
        RETURNING id INTO last_group_id;

        RAISE LOG 'New group created with group_id: %', last_group_id;

         -- Add the user to the newly created group
        INSERT INTO public.group_users (group_id, user_id, role)
        VALUES (last_group_id, NEW.id, 'admin');
    END IF;

    -- Create a new entry in user_preferences with the selected_group
    INSERT INTO public.user_preferences (id, selected_group)
    VALUES (NEW.id, last_group_id);

    RETURN NEW;
END;
$$ LANGUAGE plpgsql SECURITY DEFINER;
```

The trigger is an after insert trigger:
```js
DROP TRIGGER IF EXISTS trigger_after_insert_profiles on public.profiles;
CREATE TRIGGER trigger_after_insert_profiles
AFTER INSERT ON public.profiles
FOR EACH ROW
EXECUTE FUNCTION create_user_preferences_entry();
```

Weirdly im getting a "relation "public.group_users" does not exist" error. The table does exist and is already filled with entries. 
I can "manually" select it through a normal query.

Does anyone have an idea, what could be the problem? Im really stuck right now :/

**Comment by GaryAustin1** [comment]
Do you have any capital letters in the real table name?

**Reply by TQuantab** [reply]
No, it's all lowercase.

Heres the definition:
```js
create table
  public.group_users (
    id uuid not null default gen_random_uuid (),
    group_id uuid not null,
    user_id uuid not null,
    role text not null default 'member'::text,
    created_at timestamp with time zone null default now(),
    constraint group_users_pkey primary key (id),
    constraint group_users_group_id_fkey foreign key (group_id) references groups (id),
    constraint group_users_user_id_fkey foreign key (user_id) references auth.users (id) on update cascade on delete cascade
  ) tablespace pg_default;

create unique index group_users_group_id_idx on public.group_users using btree (group_id, user_id, role) tablespace pg_default;

create trigger on_change_update_user_metadata
after insert
or delete
or
update on group_users for each row
execute function update_user_roles ();
```

**Reply by GaryAustin1** [reply]
Did you create it in the UI?  If so the UI lets you add a space at the end (usually accidentally) but if you ask it for the definition it does not show that space.  
How did you manually access it?
try `select * from public.group_users;` in the SQL editor.  If that does not work try `select * from public."group_users ";` with a space at the end.

**Reply by TQuantab** [reply]
It was created via SQL. 
This works:
select * from public.group_users;
 
This doesnt: 
select * from public."group_users ";

**Reply by GaryAustin1** [reply]
Out of ideas.

**Reply by TQuantab** [reply]
I even tried adding entries via a separat sql query:
```js
INSERT INTO public.group_users (group_id, user_id, role)
        VALUES ('69b17c16-50a9-4e26-920c-e39c9b7a51c0', '5188383b-e43d-4c20-b4c4-052759228d0b', 'member');
```
This works

**Reply by TQuantab** [reply]
It's just really weird :/ Could it be a bug?

Does this error also appear, when one of the relations im trying to set for group_users do not exist yet?

**Reply by GaryAustin1** [reply]
You have security definer function so that should cover RLS and privileges but I would expect those to error on insert.
It is possible I'm missing the intent of the error message.  I was thinking the table, but you if you have a foreign key that does not exists yet to point to then that will be an error.

**Reply by GaryAustin1** [reply]
And I doubt it is a bug with Postgres.

**Reply by TQuantab** [reply]
Some additional Info, that could help:
A Separat function works:
```js
begin
     IF auth.uid() IS not NULL THEN 
     insert into group_users(group_id, user_id, role) values(new.id, auth.uid(), 'owner');
     end if;
     return new;
end;
```

**Reply by TQuantab** [reply]
The "flow" of the function is the following:
Auth.user gets created (via azure oauth) -> Profile gets inserted via trigger -> This function runs. Its all one process

**Reply by TQuantab** [reply]
The group definetly exists when creating the group_users. And since its an "After Insert" trigger for profiles, the auth user should also exist

**Reply by GaryAustin1** [reply]
Any chance you have other triggers going off in the process that might be the source of the error?

**Reply by TQuantab** [reply]
The only other trigger that runs, is the one i've shown above:
```js
begin
     IF auth.uid() IS not NULL THEN 
     insert into group_users(group_id, user_id, role) values(new.id, auth.uid(), 'owner');
     end if;
     return new;
end;
```

this one only runs, if auth.uid() is not null => Meaning an already registered user created a new group, then this trigger sets him as owner by default.

It has to be the Big function, when i remove 
```js
INSERT INTO public.group_users (group_id, user_id, role)
        VALUES (invite.group_id, NEW.id, 'member');
```

It runs without errors. It really seems to be, this specific line

**Reply by GaryAustin1** [reply]
You seem to have another trigger on group_users
```create trigger on_change_update_user_metadata
after insert
or delete
or
update on group_users for each row
execute function update_user_roles ();
```
What does it do?

**Reply by TQuantab** [reply]
```js
DECLARE
  _group_id UUID = COALESCE(new.group_id, old.group_id);
  _group_id_old UUID = COALESCE(old.group_id, new.group_id);
  _user_id UUID = COALESCE(new.user_id, old.user_id);
  _user_id_old UUID = COALESCE(old.user_id, new.user_id);
BEGIN
  -- Check if user_id or group_id is changed
  IF _group_id IS DISTINCT FROM _group_id_old OR _user_id IS DISTINCT FROM _user_id_old THEN
      RAISE EXCEPTION 'Changing user_id or group_id is not allowed';
  END IF;

  -- Update raw_app_meta_data in auth.users
  UPDATE auth.users
  SET raw_app_meta_data = JSONB_SET(
      raw_app_meta_data,
      '{groups}',
      JSONB_STRIP_NULLS(
        JSONB_SET(
          COALESCE(raw_app_meta_data->'groups', '{}'::JSONB),
          ARRAY[_group_id::TEXT],
          COALESCE(
            (SELECT JSONB_AGG("role")
             FROM group_users gu
             WHERE gu.group_id = _group_id
               AND gu.user_id = _user_id
            ),
            'null'::JSONB
          )
        )
      )
    )
  WHERE id = _user_id;

  -- Return null (the trigger function requires a return value)
  RETURN NULL;
END;
```

**Reply by TQuantab** [reply]
This only updates the users raw_app_meta_data to include the group info

**Reply by TQuantab** [reply]
Interesting, deactivating the trigger, removes the error :/
But i have absolutly no idea, why

**Reply by GaryAustin1** [reply, marked as answer]
I went and searched for more detail on the error and relation "table" does not exist is clearly the table not being found versus a constraint violation on a foreign key.   That normally occurs with auth initiated triggers because auth.users does not have access to the public schema or public in the search path.   You have to use security definer to get access (which you seem to do) and then either set the search path to include public in the function definition or make sure to always refer to the table by public.table versus table. 

The code you show seems to do that.  

Adding what followed was the other trigger function did not have public.

**Reply by GaryAustin1** [reply]
Yes I see it now in your 2nd trigger.  You have group_users and not public.group_users.

**Reply by GaryAustin1** [reply]
I did not get down that path initially as your main function had public.group_users.

**Reply by TQuantab** [reply]
Thank you so much for your help!!
Updating the second trigger to include "public" does the trick!

You really saved me here =D!!
